### ISSUE TITLE
`wants` documentation

### ISSUE THREAD

**Issue by horseinthesky** [wbthomason/packer.nvim#537]
### Describe the feature
I see a lot of people [use](https://github.com/folke/dot/blob/master/config/nvim/lua/plugins.lua#L181-L190) `wants` term which is not mentioned in README.

Could you please explain what it is for and how I should use it?

**Comment by akinsho** [comment]
@horseinthesky I believe the reason that `wants` isn't documented is that the plan was for it to be quite a short-lived key that solved an issue for the user that contributed it. There's some plan to refactor the way the sequencing keys work which will probably lead to it being deprecated so, it isn't being advertised or explained to avoid people using it I think, although @wbthomason might have changed his mind, but from the original discussion I don't think it was planned that it would remain available for this long. See https://github.com/wbthomason/packer.nvim/pull/264#issuecomment-805848798 for reference.

**Comment by horseinthesky** [comment]
Understood.

For not I don't quite understand how
```
use {                                      
  "hrsh7th/nvim-compe",                    
  requires = {                             
    {                                      
      "tzachar/compe-tabnine",             
      run = "./install.sh",                
      opt = true,                          
    },                                     
    { "honza/vim-snippets", opt = true },  
    {                                      
      "SirVer/ultisnips",                  
      wants = "vim-snippets",              
      opt = true,                          
    },                                     
  },                                       
  event = "InsertEnter",                   
  wants = { "compe-tabnine", "ultisnips" },
  config = function()                      
    require "config.compe"                 
  end                                      
}                                          
```
works exactly.

Removing `wants` leads to `ultisnips` is not loaded.

**Comment by akinsho** [comment]
Tbh I personally don't use the key so am not an expert on it. I've always envisioned it will change or be removed and would rather not have to tweak my config, that being said I believe the general idea around this key is that it checks if a plugin in the wants table is loaded when the plugin using the wants key is about to be loaded and if the plugin is not it then loads the plugins in the wants table, that should recursively loaded any plugins wanted by *those* plugins and so on and so for but don't know how well it all works.

**Comment by horseinthesky** [comment]
That is correct.

I decided to open the issue due to:
1) People [use](https://github.com/folke/dot/blob/master/config/nvim/lua/plugins.lua#L172-L200) the feature constantly.
2) It is not documented.
3) I don't see any other option for a plugin to load some other plugins it depends on.

**Comment by akinsho** [comment]
Will leave this for @wbthomason other than providing context not sure what else I can add. If the plan is still to change it's behaviour, rename or remove it then as was initially discussed in the issue then I advise against using it or documenting its current behaviour. The cat might be out of the bag a little and people are now using it but I think it still shouldn't be encouraged yet till the longterm plan for it is complete.

In any event I don't really have much else to add.

**Comment by atchim** [comment]
This feature doesn't seem to be working properly. Take the following setup code as example.
```lua
require'packer'.startup{
  {
    -- `nvim-web-devicons` is denpency of several other plugins.
    {'kyazdani42/nvim-web-devicons', opt = true},
    -- `nvim-tree.lua` requires `nvim-web-devicons` to be loaded first.
    {'kyazdani42/nvim-tree.lua', wants = 'nvim-web-devicons'},
  },
  {},
}
```
After running `:PackerSync` and restarting Neovim, `nvim-tree.lua` is working, but with no icons because it requires `nvim-web-devicons` to be loaded, which it is not. `:PackerStatus` can be used to check this behavior.

**Comment by wbthomason** [comment]
@atchim: Please open a new issue for the problem you're observing. Thanks!

**Comment by osamuaoki** [comment]
Here is my 2 cents:

"wants" is currently undocumented and it loads the specified package **before** the  main package in **use**.

"after" is documented oddly as it loads package **before** the  main package for **use** in  example found doc/packer.txt and README.md as :
```
   after = string or list,      -- Specifies plugins to load before this plugin.
```
This is odd and goes against code (`lua/packer/load.lua` `lua/packer/compile.lua`) and  the explanation in Sequence section.

Isn't it more like:
```
   after = string or list,      -- Specifies plugins to load after this plugin.
   wants = string or list,    -- Specifies plugins to load before this plugin.
```

**Comment by osamuaoki** [comment]
Of course, if `wants` is deprecated interface over `require` and `after`, the entry for `wants` can be skipped.  (I am new to this packer thing).

**Comment by atchim** [comment]
> @atchim: Please open a new issue for the problem you're observing. Thanks!

@wbthomason, I already created an issue for this. You can find it [here](https://github.com/wbthomason/packer.nvim/issues/615).

**Comment by nyngwang** [comment]
Nobody `wants` to care this one.

**Comment by homedirectory** [comment]
Wow, I just spent 2 hours debugging my plugin dependencies and the reason turns out to be **an undocumented option**. It is a shame that `wants` is still undocumented after almost 3 years since this issue had been opened.

Neither `after` nor `requires` let me specify *load dependencies* correctly, which is crucial when the dependencies are optional.
